File: DESC_Python/src/DE_Module/objfun.py
"""
Objective function for differential evolution optimization.

Author: Rainer Storn (original MATLAB), converted to Python3 and modified by Jiewen Zhang
Description: Implements the cost function to be minimized.
"""

import numpy as np


def objfun(FVr_temp, S_struct, S):
    """
    Objective function for differential evolution optimization.
    
    Parameters
    ----------
    FVr_temp : array-like
        Parameter vector (log10 stress drop values)
    S_struct : dict
        Contains a variety of parameters (for compatibility)
    S : dict
        Contains problem-specific data including:
        - nfreqg: frequency indices
        - freqg: frequency values
        - nlow: low frequency indices
        - evspec_DE: event spectra data
        - specall: all spectra array
        - ibin: bin indices
        - fmom: seismic moments
        - nfit: fitting indices
        - min_nspec: minimum number of spectra per bin
    
    Returns
    -------
    S_MSE : dict
        Dictionary containing:
        - I_nc: Number of constraints (0)
        - FVr_ca: Constraint values (0)
        - I_no: Number of objectives (1)
        - FVr_oa: Objective function values
    power : dict
        Dictionary containing:
        - ECSs_med: Median ECS values
        - ECSs: ECS values for all events
        - fc: Corner frequencies
    """
    
    # Extract variables
    nfreqg = S['nfreqg']
    freqg = S['freqg']
    nlow = S['nlow']
    evspec = S['evspec_DE']
    specall = S['specall']
    ibin = S['ibin']
    fmom = S['fmom']
    nfit = S['nfit']
    min_nspec = S['min_nspec']
    
    # Filter events with valid magnitude residuals
    ng = np.where(np.array([ev['qmagresid'] for ev in evspec]) <= 100)[0]
    evspec = [evspec[i] for i in ng]
    specall = specall[ng, :]
    ibin = ibin[ng]
    fmom = fmom[ng]
    
    beta = 3464
    fact = (0.42 * beta) ** 3
    
    # Optimize for matrix computation
    # Convert log10 stress drop to Pa
    dsigtarg = (10.0 ** FVr_temp[ibin]) * 1e6
    
    # Vectorized corner frequency calculation
    fc = ((dsigtarg * fact) / fmom) ** (1.0 / 3.0)
    
    # Compute predicted spectra
    freqg_mat = np.tile(freqg, (len(evspec), 1))
    fc_mat = np.tile(fc[:, np.newaxis], (1, len(freqg)))
    pred = -np.log10(1 + (freqg_mat / fc_mat) ** 2)
    
    # Calculate offset
    yoff = np.mean(specall[:, nfreqg[nlow]] - pred[:, nfreqg[nlow]], axis=1)
    
    # Calculate Empirical Coda Source spectra (ECS)
    ECSs = specall - yoff[:, np.newaxis] - pred
    
    # Calculate median ECS for each bin
    ibin_u = np.unique(ibin[~np.isnan(ibin)])
    ECSs_med = np.full((len(ibin_u), ECSs.shape[1]), np.nan)
    
    for iibin in range(len(ibin_u)):
        if np.sum(ibin == ibin_u[iibin]) >= min_nspec:
            ECSs_med[iibin, :] = np.median(ECSs[ibin == ibin_u[iibin], :], axis=0)
    
    # Calculate standard deviation across bins at fitting frequencies
    ST = np.sum(np.nanstd(ECSs_med[:, nfit], axis=0))
    
    # Store outputs
    power = {
        'ECSs_med': ECSs_med,
        'ECSs': ECSs,
        'fc': fc
    }
    
    FVr_oa = [1 / ST]

    # Output structure
    S_MSE = {
        'I_nc': 0,           # no constraints
        'FVr_ca': 0,         # no constraint array
        'I_no': 1,           # number of objectives (costs)
        'FVr_oa': FVr_oa         # objective function value
    }
    
    return S_MSE, power

File: DESC_MATLAB/src/DE_Module/objfun.m
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Function:         S_MSE= objfun(FVr_temp, S_struct)
% Author:           Rainer Storn
% Description:      Implements the cost function to be minimized.
% Parameters:       FVr_temp     (I)    Paramter vector
%                   S_Struct     (I)    Contains a variety of parameters.
%                                       For details see Rundeopt.m
% Return value:     S_MSE.I_nc   (O)    Number of constraints
%                   S_MSE.FVr_ca (O)    Constraint values. 0 means the constraints
%                                       are met. Values > 0 measure the distance
%                                       to a particular constraint.
%                   S_MSE.I_no   (O)    Number of objectives.
%                   S_MSE.FVr_oa (O)    Objective function values.
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
function [S_MSE, power]= objfun(FVr_temp, S_struct, S)

%---Objective function----------------------------------------------

%% Extract variables
nfreqg     = S.nfreqg;
freqg      = S.freqg;
nlow       = S.nlow;
evspec     = S.evspec_DE;
specall    = S.specall;
ibin       = S.ibin;
fmom       = S.fmom;
nfit       = S.nfit;
min_nspec  = S.min_nspec;

%% Organize variables

%% Compute objective function values
% Remember: here FVr_temp is log10 stress drop!!!
ng = find([evspec.qmagresid]<=100);
evspec = evspec(ng);
specall = specall(ng,:);
ibin = ibin(ng);
fmom = fmom(ng);
beta = 3464;
fact = (0.42*beta)^3;

%%% Optimize for matrix computation
tic
dsigtarg = (10.^FVr_temp(ibin)) * 1e6; % convert to Pa.
fc = (dsigtarg.*fact./fmom).^(1/3);

pred = -log10(1+(repmat(freqg,length(evspec),1)./repmat(fc',1,length(freqg))).^2);
yoff = mean(specall(:,nfreqg(nlow))-pred(:,nfreqg(nlow)),2);
ECSs = specall - repmat(yoff,1,length(freqg)) - pred;

ibin_u = unique(ibin(~isnan(ibin)));
ECSs_med = nan*ones(length(ibin_u),size(ECSs,2));
for iibin = 1:length(ibin_u)
    if(sum(ibin==ibin_u(iibin))>=min_nspec)
        ECSs_med(iibin,:) = median(ECSs(ibin==ibin_u(iibin),:),1);
    end
end

ST = sum(nanstd(ECSs_med(:,nfit),[],1));

power.ECSs_med = ECSs_med;
power.ECSs = ECSs;
power.fc = fc;

%% Output
% ----strategy to put everything into a cost function------------
S_MSE.I_nc      = 0;%no constraints
S_MSE.FVr_ca    = 0;%no constraint array
S_MSE.I_no      = 1;%number of objectives (costs)
S_MSE.FVr_oa(1) = 1/ST;
